Adicionar testes para algoritmos pós-quânticos (Keygen, Sign e Verify)

bifee opened the pull request:
### O que foi feito:
- Implementados testes funcionais para os algoritmos pós-quânticos, cobrindo as operações de:
  - Geração de chaves (Keygen)
  - Assinatura de mensagens (Sign)
  - Verificação de assinaturas (Verify)
- Cada teste executa os algoritmos dentro de um período fixo de tempo (3 segundos por padrão).
- Métricas de desempenho adicionais foram calculadas para cada algoritmo.

### Algoritmos testados:
- ML-DSA-44
- ML-DSA-65
- ML-DSA-87
- cross-rsdp-128-small
- cross-rsdp-128-fast
- cross-rsdp-192-small
- cross-rsdp-256-small

### Métricas calculadas:
1. **Número de iterações**: Quantidade de operações executadas dentro do tempo fixo.
2. **Tempo médio por operação (µs/op)**: Tempo médio em microssegundos para completar cada operação.
3. **Ciclos médios por operação (cycles/op)**: Número médio de ciclos da CPU consumidos por operação, usando a biblioteca `gotsc`.
4. **Operações por segundo (ops/sec)**: Taxa de operações realizadas por segundo.

### Como rodar os testes:
1. Certifique-se de que as dependências estão instaladas:
   - Biblioteca `liboqs-go` para algoritmos pós-quânticos.
   - Biblioteca `gotsc` para medição de ciclos da CPU.

2. Execute os testes com o comando:
```
go test -v -duration=5s
```
3. Exemplos de saida
```
TESTING KEYGEN - Algorithm: ML-DSA-65 | Iterations: 18701 | Mean CPU Cycles: 2196813163 | Mean Time: 1157334 µs | Operations/S: 6233.666667 s
```

### Observações:
- O tempo de execução dos testes é configurável via flag `-duration` (padrão: 3 segundos).
- Resultados podem variar dependendo do hardware e da carga do sistema.

bifee committed:
Adicionados testes para algoritmos pós-quânticos (Keygen, Sign e Verify)

### O que foi feito:
- Implementados testes funcionais para os algoritmos pós-quânticos, cobrindo as operações de:
  - Geração de chaves (Keygen)
  - Assinatura de mensagens (Sign)
  - Verificação de assinaturas (Verify)
- Cada teste executa os algoritmos dentro de um período fixo de tempo (3 segundos por padrão).
- Métricas de desempenho adicionais foram calculadas para cada algoritmo.

### Algoritmos testados:
- ML-DSA-44
- ML-DSA-65
- ML-DSA-87
- cross-rsdp-128-small
- cross-rsdp-128-fast
- cross-rsdp-192-small
- cross-rsdp-256-small

### Métricas calculadas:
1. **Número de iterações**: Quantidade de operações executadas dentro do tempo fixo.
2. **Tempo médio por operação (µs/op)**: Tempo médio em microssegundos para completar cada operação.
3. **Ciclos médios por operação (cycles/op)**: Número médio de ciclos da CPU consumidos por operação, usando a biblioteca `gotsc`.
4. **Operações por segundo (ops/sec)**: Taxa de operações realizadas por segundo.

### Como rodar os testes:
1. Certifique-se de que as dependências estão instaladas:
   - Biblioteca `liboqs-go` para algoritmos pós-quânticos.
   - Biblioteca `gotsc` para medição de ciclos da CPU.

2. Execute os testes com o comando:
```bash
go test -v -duration=5s

3. Exemplos de saida
'''bash
TESTING KEYGEN - Algorithm: ML-DSA-65 | Iterations: 18701 | Mean CPU Cycles: 2196813163 | Mean Time: 1157334 µs | Operations/S: 6233.666667 s

### Observações:
- O tempo de execução dos testes é configurável via flag `-duration` (padrão: 3 segundos).
- Resultados podem variar dependendo do hardware e da carga do sistema.

alexandreboutrik reviewed (changes requested):
Bom dia / Boa noite Sr @bifee 

No módulo [hybrid_sig](https://github.com/PQC-Group-UTFPR/go-std/blob/master/src/crypto/hybrid/hybrid_sig.go), optamos por utilizar a variante RSDPG ao invés da RSDP do CROSS, devido ao [tamanho das chaves públicas](https://github.com/rtjk/CROSS-PQClean/blob/main/generate/parameter_sets.csv). Se puder adicionar o sufixo `g` nas strings `cross-*` do array `algorithms` (ex: cross-rsdpg-128-small, ...).

Além disso, no relatório trimestral, como planejamos fazer um benchmark de comparação entre a liboqs e liboqs-go para o algoritmo Mayo (strings: Mayo-1, Mayo-2, Mayo-3, Mayo-5), talvez seja interessante adicioná-lo também. Dessa forma podemos fechar a Issue #3 .

Caso optemos por isso, acredito que seja melhor gerar um `hash` aleatório (em vez de definir um valor constante, linha 37) para manter uma implementação alinhada ao comportamento do [speed_sig.c](https://github.com/open-quantum-safe/liboqs/blob/main/tests/speed_sig.c) (linha 61) da liboqs em C.

O que acha Prof @PQC-Group-UTFPR ?

PQC-Group-UTFPR commented:
Perfeito, @alexandreboutrik. Obrigado pela revisão

Você conseguiria orientar o @Yuuto-Senpai para ele fazer estas modificações?

- [x] Trocar sufixo nos parâmetros do CROSS
- [x] Adicionar o MAYO nos parâmetros mencionados
- [x] (Sugestão) Mover este código para `src/crypto/hybrid/`. Pode ser? Prefiro manter nossos códigos em um só lugar.

Assim ajudaria a ele contribuir neste repositório.

Na questão do "hash" citada ali podemos manter assim. OQS_Randombytes vai ter um comportamento semelhante e não é contabilizado no tempo. Então pode manter. Mas caso queiram, use o pacote [rand](https://pkg.go.dev/crypto/rand) (tem exemplo lá). Só não o façam dentro do timer para não contabilizar o tempo (ou seja, conforme é feito em C).

Atenciosamente,
Alexandre Giron

PQC-Group-UTFPR reviewed (changes requested):
Conforme comentário anterior.

gabrielhbueno committed:
Update pqc_benchmark_test.go

O que foi feito!

1.Adição dos algoritmos MAYO no array algorithms.

2.Troca dos sufixos nos parâmetros CROSS para usar rsdpg no lugar de rsdp.

3.Geração de hash aleatório por meio da função generateRandomHash().

alexandreboutrik commented:
Boa tarde Sr @Yuuto-Senpai 

Algumas observações:
1. Para os próximos commits, se puder seguir o padrão da [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
2. O comentário da linha 28 é desnecessária: "// conforme Prof e Boutrik". Por exemplo, quem for ler o código futuramente não entenderá o contexto deste pull request ou do que conversamos por fora. Seria bom removê-lo.
3. O comentário da função `generateRandomHash` ficou depois da função, o correto seria acima dela, para tornar o comentário [Doc comment](https://go.dev/doc/comment).
4. Na linha 27 do diff, você removeu um comentário, e na linha 16, você adicionou um quebra linha. Não são detalhes muito importantes, porém seria bom evitar esse tipo de coisa em commits futuros para não poluir o diff.
5. Na linha 48, acredito que seja melhor seguir o padrão de utilizar o `t.Fatalf` para tratar de erros, ao invés de `panic(fmt.Printf)`.
6. Sobre a geração do hash, é melhor gerar ele uma única vez (ex: de forma global) ao invés de gerar ele várias vezes localmente nas funções. Seria bom trocar o tamanho do hash, de 32, para 50, como está no [speed_sig.c](https://github.com/open-quantum-safe/liboqs/blob/main/tests/speed_sig.c) (linha 42) da liboqs em C. O tamanho da mensagem pode talvez afetar na performance dos Signs.

gabrielhbueno committed:
fix(pqc_benchmark_test.go): ajustes nos comentários, hash e tratamento de erros

- Removido comentário desnecessário na linha 28.
- Posicionado comentário Doc Comment acima da função `generateRandomHash`.
- Ajustado o uso de `t.Fatalf` na linha 48 para tratar erros de forma consistente.
- Geração do hash ajustada para ocorrer globalmente, alinhando com o padrão do `speed_sig.c`.
- Alterado o tamanho do hash de 32 bytes para 50 bytes.
- Removidas quebras de linha e ajustes não relacionados para evitar poluição do diff.

PQC-Group-UTFPR reviewed (approved):
Faltou mover para a pasta `hybrid`. Mas isso é fácil de resolver depois. Vamos seguir com o merge e com o benchmark.
